language-plutus-core/src/Language/PlutusCore/Pretty/Refined.hs

```diff
@@ -0,0 +1,160 @@
+-- | A "refined" ("readable", "Agda-like") way to pretty-print PLC entities.
+
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+module Language.PlutusCore.Pretty.Refined
+    ( RenderContext (..)
+    , PrettyConfigRefined (..)
+    , PrettyRefinedBy
+    ) where
+
+import           Language.PlutusCore.Lexer.Type     hiding (name)
+import           Language.PlutusCore.Type
+import           PlutusPrelude
+
+import           Data.Text.Prettyprint.Doc.Internal (enclose)
+
+-- I wanted to take the semantics approach to pretty-printing here instead of simply deciding
+-- where to put parens dependening on precedence and associativity. That was a mistake, because
+-- the precedence-and-associativity approach is not ad hoc and can be applied in other cases as well.
+-- At some point we may even decide to incorporate precedence-and-associativity into the 'PrettyBy'
+-- class in the style of 'showPrec' to get nice pretty-printing output across different pretty-printing
+-- strategies and data types, but the current version seems to work alright, so it's fine for now.
+
+data DocStructure
```

Comment:
I decided not to document it, because
1. I really should have used the precedence-and-associativity approach, so I kinda plan to change it in future
2. That's just an internal module that doesn't export any of this stuff except `RenderContext`.
If you think this should be documented still, that's no problem, I can do that.

Reply:
> I sympathise with 1, but in my experience these things tend to hang around longer than you would like ;)

True.

Okay, I'll add some docs.

Reply:
The pain of writing docs with the intention of throwing them away drove me to implementation the precedence-and-associativity approach.